Lib/test/test_array.py

```diff
@@ -40,6 +40,12 @@ def test_bad_constructor(self):
         self.assertRaises(TypeError, array.array, 'xx')
         self.assertRaises(ValueError, array.array, 'x')
 
+    @support.cpython_only
+    def test_uninitialised_new(self):
+        # Prevent heap types from being created uninitialised (bpo-43916)
+        tp = type(iter(array.array('I')))
+        self.assertRaises(TypeError, tp)
```

Comment:
If it would expect, the constructor would expect an argument:

```suggestion
        my_array = array.array('I')
        arrayiterator = type(iter(my_array))
        self.assertRaises(TypeError, arrayiterator, my_array)
```

Reply:
Hm, I don't think so.

```
$ python3.10                  
>>> import array
>>> tp = type(iter(array.array('I')))
>>> tp()
<array.arrayiterator object at 0x10f934140>
>>> tp([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: array.arrayiterator() takes no arguments
>>> 
```

Reply:
Ditto for the similar suggestions.

Reply:
On what do you want to iterate if there is no argument?

Obviously, today, it takes no argument... because the type doesn't implement tp_new nor tp_init ;-)

Reply:
Does it really have anything to say? We hit the right exception anyway, AFAICS.

```
$ ./python.exe 
Python 3.10.0a7+ (heads/add-disallow-instantiation-flag:71439d0de8, Apr 30 2021, 15:22:20) [Clang 12.0.5 (clang-1205.0.22.9)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import array
>>> tp = type(iter(array.array('I')))
>>> tp()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: cannot create 'array.arrayiterator' instances
>>> tp([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: cannot create 'array.arrayiterator' instances
```

Reply:
My concern can be addressed later by adding an helper function which checks the error message.

Reply:
I insist on this issue because test_sys had a test which still passed when I modified sys.flags type to allow instantiation. I had to fix the test: https://github.com/python/cpython/commit/3bb09947ec4837de75532e21dd4bd25db0a1f1b7

Reply:
Ah, I got it now. I misunderstood you first. Yes, that makes sense. We'll fix it when we've got a helper in place.